src/simularium/FrameRecorder.ts

```diff
@@ -0,0 +1,144 @@
+import { ArrayBufferTarget, Muxer } from "mp4-muxer";
+import { DEFAULT_FRAME_RATE } from "../constants";
+
+/**
+ * Records frames to an MP4 file using the WebCodecs API.
+ * Resulting file is passed to the front end implementation
+ * for download/use, via the handleFile() callback.
+ *
+ *
+ * Note that the VideoCodecs API is unavailable in some browsers, including Firefox,
+ * as of 2/6/2024. Viewport will not call these methods on firefox.
+ *
+ */
+export class FrameRecorder {
+    private getCanvas: () => HTMLCanvasElement | null;
+    private handleFile: (videoBlob: Blob) => void;
+    private encoder: VideoEncoder | null;
+    private muxer?: Muxer<ArrayBufferTarget>;
+    public isRecording: boolean;
+    private frameIndex: number;
+    public supportedBrowser: boolean;
+    private frameRate: number;
+
+    constructor(
+        getCanvas: () => HTMLCanvasElement | null,
+        handleFile: (videoBlob: Blob) => void
+    ) {
+        this.getCanvas = getCanvas;
+        this.handleFile = handleFile;
+        this.encoder = null;
+        this.isRecording = false;
+        this.frameIndex = 0;
+        this.supportedBrowser = !/firefox|fxios/i.test(navigator.userAgent);
+        this.frameRate = DEFAULT_FRAME_RATE;
+    }
+
+    private async setup(): Promise<void> {
+        const canvas = this.getCanvas();
+        if (canvas) {
+            try {
+                // VideoEncoder sends chunks of frame data to the muxer.
+                // Previously made one encoder in the constructor but
+                // making a new one during setup() prevents a bug where
+                // frames returned blank (stale canvas reference?)
+                this.encoder = new VideoEncoder({
+                    output: (chunk, meta) => {
+                        if (this.isRecording && this.muxer) {
+                            this.muxer?.addVideoChunk(chunk, meta);
+                        }
+                    },
+                    error: (error) => {
+                        console.error("Encoder error:", error);
+                    },
+                });
+                const config: VideoEncoderConfig = {
+                    codec: "avc1.420028",
+                    width: canvas.width,
+                    height: canvas.height,
+                };
+                const { supported, config: supportedConfig } =
+                    await VideoEncoder.isConfigSupported(config);
+                if (supported && supportedConfig) {
+                    this.encoder.configure(config);
+                } else {
+                    throw new Error("Unsupported video encoder configuration");
```

Comment:
Same question here, have you seen this error happen and when?

Reply:
In regard to this and above, tbh I was including error handling where the docs/examples seemed to have it. Caniuse says `avc` is supported almost everywhere. Except Opera mini and some very old browser versions